### PR TITLE
Receipt design changes FormRecognizer

### PR DESCRIPTION
**Pull request by samvaity**

### CODE REVIEW

File: sdk/formrecognizer/azure-ai-formrecognizer/src/samples/java/com/azure/ai/formrecognizer/StronglyTypedRecognizedForm.java [hunk 1]
```diff
@@ -41,36 +41,40 @@ public static void main(final String[] args) {
...
+            System.out.printf("Merchant Name: %s, confidence: %.2f%n", usReceipt.getMerchantName().getFieldValue(),
                 usReceipt.getMerchantName().getConfidence());
             System.out.printf("Merchant Address: %s, confidence: %.2f%n",
-                usReceipt.getMerchantAddress().getValue(),
+                usReceipt.getMerchantAddress().getFieldValue(),
                 usReceipt.getMerchantAddress().getConfidence());
+            // It will have to be this way?
+            // System.out.printf("Merchant Address: %s, confidence: %.2f%n",
+            //     usReceipt.getMerchantAddress().getFieldValue().asString(),
+            //     usReceipt.getMerchantAddress().getConfidence());
```
Comment (samvaity) [inline sdk/formrecognizer/azure-ai-formrecognizer/src/samples/java/com/azure/ai/formrecognizer/StronglyTypedRecognizedForm.java:52]: @srnagar Confirming if this how we would be using the `FormField<T>` for strongly typed examples?

File: sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/Transforms.java [hunk 2]
```diff
@@ -244,39 +244,57 @@ private Transforms() {
         FormField<?> value;
         switch (fieldValue.getType()) {
             case PHONE_NUMBER:
-                value = new FormField<>(setDefaultConfidenceValue(fieldValue.getConfidence()), labelText,
-                    key, fieldValue.getValuePhoneNumber(), valueText, FieldValueType.PHONE_NUMBER);
+                value = new FormField<>(key, labelText, valueText, new com.azure.ai.formrecognizer.models.FieldValue(FieldValueType.PHONE_NUMBER).
```
Comment (mssfang) [inline sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/Transforms.java:247]: nit: dot might need to be at beginning of line for style consistency.

File: sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/Transforms.java [hunk 3]
```diff
@@ -244,39 +244,57 @@ private Transforms() {
...
+                setFormFieldPhoneNumber(fieldValue.getValuePhoneNumber()), setDefaultConfidenceValue(fieldValue.getConfidence())
+                );
                 break;
             case STRING:
-                value = new FormField<>(setDefaultConfidenceValue(fieldValue.getConfidence()), labelText,
-                    key, fieldValue.getValueString(), valueText, FieldValueType.STRING);
+                value = new FormField<>(key, labelText, valueText, new com.azure.ai.formrecognizer.models.FieldValue(FieldValueType.STRING)
```
Comment (mssfang) [inline sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/Transforms.java:252]: nit: com.azure.ai.formrecognizer.models.FieldValue can be in import list at the beginning of file

File: sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/models/FormField.java [hunk 4]
```diff
@@ -14,28 +14,25 @@
...
+     * @param labelData The text, bounding box, and field elements for the field label.
      * @param valueData The text, bounding box, and field elements for the field value.
-     * @param valueType The type of the value of the recognized field.
+     * @param value The value of the recognized field.
+     * @param confidence The confidence of the recognized field.
      */
-    public FormField(final float confidence, final FieldData labelData, final String name, final T value,
-        final FieldData valueData, FieldValueType valueType) {
+    public FormField(final String name, final FieldData labelData, final FieldData valueData,
+        final FieldValue value, final float confidence) {
```
Comment (mssfang) [inline sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/models/FormField.java:30]: no longer takes a FieldValueType?

File: sdk/formrecognizer/azure-ai-formrecognizer/src/samples/java/com/azure/ai/formrecognizer/AdvancedDiffLabeledUnlabeledDataAsync.java [hunk 5]
```diff
@@ -168,7 +169,8 @@ public static void main(String[] args) throws IOException {
                     .filter(formFieldEntry -> "Vendor Name:".equals(formFieldEntry.getValue().getLabelData().getText()))
                     .findAny()
                     .ifPresent(formFieldEntry ->
-                        System.out.printf("The Vendor name is: %s%n", formFieldEntry.getValue()));
+                        System.out.printf("The Vendor name is: %s%n", formFieldEntry.getValue()
```
Comment (mssfang) [inline sdk/formrecognizer/azure-ai-formrecognizer/src/samples/java/com/azure/ai/formrecognizer/AdvancedDiffLabeledUnlabeledDataAsync.java:172]: feel strange but it is what it is.

File: sdk/formrecognizer/azure-ai-formrecognizer/src/samples/java/com/azure/ai/formrecognizer/Receipt.java [hunk 6]
```diff
@@ -357,4 +356,6 @@ public ReceiptItem(final FormField<String> name, final FormField<Float> quantity
             return totalPrice;
         }
     }
+
+
```
Comment (mssfang) [inline sdk/formrecognizer/azure-ai-formrecognizer/src/samples/java/com/azure/ai/formrecognizer/Receipt.java:360]: nit: extra line

File: sdk/formrecognizer/azure-ai-formrecognizer/src/test/java/com/azure/ai/formrecognizer/FieldValueExtensionMethodTest.java [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+    @Test
+    public void toMapFromMap() {
+        Map<String, FormField<?>> inputMap = new HashMap<String, FormField<?>>() {
+            {
+                put("key", new FormField<>(null, null, null, null, 0));
+            }
+        };
+        Map<String, FormField<?>> actualList = new FormField<>(null, null, null,
+            new FieldValue(FieldValueType.MAP).setFormFieldMap(inputMap), 0).getValue().asMap();
+        assertEquals(inputMap, actualList);
```
Comment (mssfang) [inline sdk/formrecognizer/azure-ai-formrecognizer/src/test/java/com/azure/ai/formrecognizer/FieldValueExtensionMethodTest.java:183]: assertTrue(inputMap.equals(actualList));
https://www.baeldung.com/java-compare-hashmaps

File: sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/Transforms.java (hunk 3 above)
Comment (samvaity) [reply sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/Transforms.java:252]: Can't..have FieldValue in implementation.models too.

File: sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/models/FormField.java (hunk 4 above)
Comment (samvaity) [reply sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/models/FormField.java:30]: It is in the `FieldValue` model.

File: sdk/formrecognizer/azure-ai-formrecognizer/src/samples/java/com/azure/ai/formrecognizer/StronglyTypedRecognizedForm.java (hunk 1 above)
Comment (srnagar) [reply sdk/formrecognizer/azure-ai-formrecognizer/src/samples/java/com/azure/ai/formrecognizer/StronglyTypedRecognizedForm.java:52]: I don't think the generic type `<T>` is useful if the user still has to do `getFieldValue().asString()`. I think we are trying to make `FormField` work both ways and is causing this issue.  

Have you considered the approach below? It's not ideal but a new type that's defined once can reduce the burden on the user. They'll not have to repeatedly call with `getValue().asType()` methods which is going to be accessed a lot more frequently.

```java

// Receipt type
public final class Receipt {
    private ReceiptType receiptType;
    private TypedFormField<String> merchantAddress; // return a different form field type
    
    public Receipt(RecognizedForm recognizedForm) {

       for (Map.Entry<String, FormField<?>> entry : recognizedForm.getFields().entrySet()) {
            String key = entry.getKey();
            FormField<?> fieldValue = entry.getValue();
            switch (key) {
                case "MerchantAddress":
                    merchantAddress = new TypedFormField(fieldValue, String.class);
                    break;
                
                default:
                    break;
            }
        }
        
    }

    // getter and setters here
    public TypedFormField<String> getMerchantAddress() {
         return merchantAddress;
    }
}

-------------------------------------------

// Strongly typed form field
public class TypedFormField<T> { // with a better class name

    private final FormField formField;
    private final Class<T> type;

    public TypedFormField(FormField formField, Class<T> type) {
        this.formField = formField;
        this.type = type;
    }

    public T getValue() {
        if (formField.getValueType() == FieldValueType.STRING && type.getClass().equals(String.class)) {
            return formField.getValue().asDate();
        }
        throw new IllegalStateException("Type mismatch error");
    }
   
    public float getConfidence() {
        return this.formField.getConfidence();
    }
 
}
-------------------------------------------
// User code
Receipt usReceipt = new Receipt(recognizedForm);
String address = usReceipt.getMerchantAddress().getValue()
float confidence = usReceipt.getMerchantAddress().getConfidence();
```

**Comment by samvaity** [comment]
>// Strongly typed form field
public class TypedFormField<T> { // with a better class name

    private final FormField formField;
    private final Class<T> type;
}

@srnagar This would mean that it will be a `TypedFormField<String>` for a phone number. Should it be fine to support that as phone number or would it have a  semantic difference?

File: sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/models/FieldValue.java [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+    private String formFieldString;
+    private String formFieldPhoneNumber;
+
+    /**
+     * Constructs a FieldValue object
+     *
+     * @param type The type of the field.
+     */
+    public FieldValue(final FieldValueType type) {
+        this.type = type;
```
[the remark this answers is not in the document]
Comment (samvaity) [reply sdk/formrecognizer/azure-ai-formrecognizer/src/main/java/com/azure/ai/formrecognizer/models/FieldValue.java:46]: `type` is required in swagger. Should we still check here then?
https://github.com/Azure/azure-rest-api-specs/blob/master/specification/cognitiveservices/data-plane/FormRecognizer/preview/v2.0/FormRecognizer.json#L1553

File: sdk/formrecognizer/azure-ai-formrecognizer/src/test/java/com/azure/ai/formrecognizer/FormRecognizerClientTestBase.java [hunk 9]
```diff
@@ -795,15 +796,15 @@ static void validateMultipageReceiptData(List<RecognizedForm> recognizedReceipts
...
         assertEquals(3, receiptPage3.getFormPageRange().getFirstPageNumber());
         assertEquals(3, receiptPage3.getFormPageRange().getLastPageNumber());
-        Map<String, FormField<?>> receiptPage3Fields = receiptPage3.getFields();
-        assertEquals(EXPECTED_MULTIPAGE_ADDRESS_VALUE, FieldValueType.STRING.cast(receiptPage3Fields.get("MerchantAddress")));
-        assertEquals("Frodo Baggins", FieldValueType.STRING.cast(receiptPage3Fields.get("MerchantName")));
-        assertEquals(EXPECTED_MULTIPAGE_PHONE_NUMBER_VALUE, FieldValueType.PHONE_NUMBER.cast(receiptPage3Fields.get("MerchantPhoneNumber")));
-        assertNotNull(receiptPage3Fields.get("Total").getValue());
+        Map<String, FormField> receiptPage3Fields = receiptPage3.getFields();
+        assertEquals(EXPECTED_MULTIPAGE_ADDRESS_VALUE, receiptPage3Fields.get("MerchantAddress").getValue().asString());
+        assertEquals("Frodo Baggins", receiptPage3Fields.get("MerchantName").getValue().asString());
```
Comment (srnagar) [inline sdk/formrecognizer/azure-ai-formrecognizer/src/test/java/com/azure/ai/formrecognizer/FormRecognizerClientTestBase.java:801]: Is this merchant selling the ring?

**Pull request merged**